### DISCUSSION TITLE
Tutorial on creating own robot xml + rhino file?

### DISCUSSION THREAD

**Discussion by yijiangh** [visose/Robots#76] [Q&A] [answered]
Hi @visose ,

Thank you for creating such a wonderful package! I am wondering if it's possible to document a bit more on creating one's own robotcell (xml + rhino file)?

From my observation (snapshot below on the Bartlett robotcell), it seems like the xml and the geometry data in the rhino file is linked by layer's name, e.g. the layer in rhino should be `RobotArm.<manufacturer>.<robot model>` and the xml file and rhino file are linked by the same file name, right?

![image](https://user-images.githubusercontent.com/11905844/41570808-525a22e0-733e-11e8-8222-7d5d51acf06c.png)

Apart from this, I am confused in the following three parts:

(1) In the xml file, for each joint we have `<Revolute number="1" a ="25" d ="400" minrange = "-170" maxrange ="170" maxspeed ="360"/>`. What do `a` and `d`'s value mean?

(2) Which parameter in the xml file specify the current robot's joint pose in the rhino file? (like the robot pose in the picture above, it is definitely not the all-zero configuration)

(3) If I want to configure a 6-axis robot on a linear track, how should I do this? Should I create a new robotcell entry in the xml file to specify a 6+1=7 axis robot, or should I just treat the linear track as the second robot system, and input it, together with the 6-axis robot, in an ordered list to the `create program` component?

Finally, if I'd love to use this package in an academic project, what would be the preferrable way to cite this package?

Thank you so much,
Yijiang

**Comment by visose** [comment]
Hi yijiangh, if you tell me the robot models and track model you have, and name of your institution, I can prepare a set of .xml and .3dm files for you. This will also help you understand how the xml works in case you need to modify it later. I agree that this should be documented somewhere.

1 - These are the Denavit–Hartenberg parameters (https://en.wikipedia.org/wiki/Denavit–Hartenberg_parameters)
2 - The DH parameters define the relative positions of the joints (as a chain).

3- In the Bartlett .xml file you'll see a how to insert a track. You should have one <RobotCell> per robot controller. If the track and robot are connected to the same controller, you'll have two mechanisms inside one RobotCell, you'll probably also have both of them inside the same Mechanical Group.

You can cite this as: Robots plugin for Grasshopper (https://github.com/visose/Robots)

**Comment by yijiangh** [comment]
Hi Vicente,

Thank you for your explanation. I am a grad student at MIT and I am trying to use your package as a simulation tool. The robot is an [IRB 6640-185-280 ](https://search-ext.abb.com/library/Download.aspx?DocumentID=ROB0001EN&LanguageCode=en&DocumentPartId=&Action=Launch) and the track model is the UCLA KUKA model that I get from the KUKA|PRC package. MIT does not currently own these giant machines but I'd love to work with them visually.

The only thing that I am having trouble with is the DH parameters in the xml file - do you have some documentation on how do you generate them? 

I've set up a draft version of the xml and rhino file (attached here: [Robots-MIT_sim.zip](https://github.com/visose/Robots/files/2121127/Robots-MIT_sim.zip)), where most of the data has been filled up except the DH parameters. Could you please help me complete them? For your convenience, I create a sublayer called `joint_points` to mark the joint positions (see snapshot below).

![image](https://user-images.githubusercontent.com/11905844/41684723-5905e088-74ac-11e8-9088-6c668700a0f9.png)

I am expecting the finished setup to be the same as the following:

![image](https://user-images.githubusercontent.com/11905844/41684766-7b2f9352-74ac-11e8-9b12-c0b43ea2920e.png)

Thank you so much for your patience and help!

**Comment by visose** [comment, marked as answer]
Hi, attached is the working modified library. I added some comments in the .xml.
The DH parameters you can get them from robot dimensions (see in comments).
It's a bit weird that you're using a KUKA track with an ABB robot. I would try to find out the actual track model for better naming conventions (rather than naming it UCLA Track) but it will work anyway.
In the Rhino file a robot should have only 7 meshes (base + 6 axes), each in the apropiate layer. I joined the different mesh parts together and applied some mesh decimation (_reducemesh command).

[MIT_sim.zip](https://github.com/visose/Robots/files/2123816/MIT_sim.zip)

**Comment by yijiangh** [comment]
Thank you very much! The example definitely helps me better understand how these things should be set up. With this, I will close the issue.

**Comment by jeojeda** [comment]
Hello @yijiangh @visose 
I am trying to follow your work to see if I am able to add an old robot that we have in TU Darmstadt. It is a welding robot COMAU SMART NM 16-3.1
Do you recomend any file I can use as a reference to adjust and be able to generate a new library?

<img src="https://raw.githubusercontent.com/jeojeda/Robot_Comau/main/SMART%20NM%2016-3.1.JPG">

Thank you very much
Juan

**Reply by visose** [reply]
This software doesn't support COMAU robots, but if you only care for the simulation, you can probably modify an existing KUKA or ABB robot. It looks pretty similar to a KUKA KR 30, so you can start and adjust the relevant parameters. The simulation will probably work, but not the generated program.

**Reply by jeojeda** [reply]
thanks @visose. For now it is ok if I can only get the simulation.

**Reply by Finnglish111** [reply]
Hi @jeojeda,

Did you get the simulation up and running for your Comau?
I was wondering if you would mind sharing your 3dm and xml files as a reference, as I am keen to do the same for our robot.

Thanks!

**Reply by jeojeda** [reply]
Hi @Finnglish111, try these files https://github.com/jeojeda/robots_comau/tree/main
I am using a Kuka xml file as base with all the values for a Comau robot

**Comment by Lorenzo-ctrl** [comment]
Hi, I'm trying to import abb crb 15000 on grasshopper. I've a problem with the dh parameters, the 6th axis doesn't work. Will it be added on robots library? @visose

**Reply by visose** [reply]
Hi @Lorenzo-ctrl, there is currently no plan to add this robot. If you or your organization are interested in sponsoring this work, I can study the feasibility of integrating it into the plugin and give you an estimate based on the expected amount of work.

**Comment by a-tiwari17** [comment]
Hi @visose , 

[Braccio.zip](https://github.com/visose/Robots/files/9862223/Braccio.zip)

I am trying to make a library for the Arduino Braccio, a small 5 axis robot, for prototyping. I followed the discussion above and attempted to put together an xml file, attached above. While I still am trying to get a .3dm file ready, I was wondering if this goes in the right direction? I am a bit confused in initialising this step 
Base x="0.000" y="0.000" z="0.000" q1="1.000" q2="0.000" q3="0.000" q4="0.000"/
because the DH parameters I found seem to have a q5 parameter as well?

**Reply by a-tiwari17** [reply]
![image](https://user-images.githubusercontent.com/101123655/197838287-da630aa6-ce26-45b9-84d4-6542b2236b78.png)

Also - I get the error "Manufacturer Arduino is not valid", when I tried using a WIP .3dm file with this xml file. I was wondering if there is something more to be done to get it to work, apart from preparing the .3dm and .xml files?

**Reply by visose** [reply]
Hi, this software doesn't support 5 axis robots, so this would have to be implemented as a new feature. You will be able to make it work via just an .xml file.

The q1, q2, q3, q4 values correspond to quaternion values, a way to represent a 3D rotation. It's not related to the robot axes.

**Reply by a-tiwari17** [reply]
Thanks for the reply! I will try what I can, and follow up shortly

**Reply by visose** [reply]
Oops, that was supposed to say "You will **not** be able to make it work via just an .xml file."

That was a very misleading typo, sorry.

**Reply by a-tiwari17** [reply]
Oh no :') 
Still, I will get back soon - I am unsure now if the Braccio is 5 or 6 axis - there seem to be both options. Maybe if there is a 6 axis version... !
